[novafinedust] Optimizations on access to the serial port

t2000 opened the pull request:
- retry logic if port does not yet exist on startup
- do not write sleep command on shutdown if port has issues
- no not register data listener on port but wait for data instead to be
  compatible with RFC2217 serial over network implementation (see https://github.com/openhab/openhab-core/issues/2170)
- ignore all buffered data from device during initialization to get the
  device into a defined state

Signed-off-by: Stefan Triller <github@stefantriller.de>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

t2000 commented:
This PR makes the usage of the binding MUCH more reliable, i.e. I do not fear restarting my openHAB installation where I had to pray whether or not the device is still working afterwards.
Also the RC2217 implementation (via ser2net) for the serial port works much better as the direct access to the port.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -78,33 +86,40 @@ public boolean initialize(WorkMode mode, Duration interval)
             throws PortInUseException, TooManyListenersException, IOException, UnsupportedCommOperationException {
         boolean initSuccessful = true;
 
+        logger.debug("Initializing with mode={}, interval={}", mode, interval);
```
I suggest changing the logging level of most of these to `trace` instead.

My general rule for logging levels is:
* trace - used for normal expected execution paths
* debug - used for unexepected execution paths, but well within normal operation
* info - used for notable points in an execution path, like a milestone. (in openhab we try reserve this logging level for the core, so bindings should rarely ever call this.)
* warn - used for notable unexpected execution paths that a regular user (not just a developer) should be notified of. Warnings should be used to indicate that something not-normal occurred and user intervention is required to resolve. Warnings do not indicate a failure to operate merely an abnormal condition of operation that can still be handled by the binding. Failures in binding operation should be indicated by changing the thing status to offline.
* error - used to indicate catastrophic program failure. This should be used to indicate a catastrophic failure in openhab's ability to operate. A failure in a binding would never cause openhab as a whole to fail so a failure in a bindings should never log an error. Instead that failure should be indicated by changing the thing status.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -252,64 +284,67 @@ public void requestSensorData() throws IOException {
+        if (localScheduler != null) {
+            Future<?> sleepJob = null;
+            try {
+                sleepJob = localScheduler.submit(new Runnable() {
```
Please use a lambda here instead.

t2000 commented:
Thanks for your review. I have moved most logging to TRACE and changed the scheduling of the sleepjob to use a lambda.

I would appreciate if you can have another look. Thanks!

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -116,7 +131,12 @@ public boolean initialize(WorkMode mode, Duration interval)
+        try {
+            write(commandData);
+        } catch (IOException ioex) {
+            logger.debug("Got an exception while writing a command, will not try to fetch a reply for it.");
```
You should include the exception message in the log so that it is easier to know what went wrong.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -116,59 +120,88 @@ public void initialize() {
+                    try {
+                        localCommunicator.readSensorData();
+                    } catch (IOException e) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                                "Cannot query data from device");
```
You should include the exception message in the status detail so that users have a better idea of what went wrong.

cpmeister reviewed (changes requested):
Only a few last changes I'd like.

t2000 commented:
@cpmeister Thank you very much, I have integrated your changes.
